[dynamo][hop] Introduce Local Map HOP

xmfan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #163133
* #162702
* __->__ #161458

Can't actually deploy it because of: https://github.com/pytorch/pytorch/issues/161456

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update

[ghstack-poisoned]

wconstab commented on test/higher_order_ops/test_local_map.py:
```diff
@@ -0,0 +1,123 @@
+    return context_parallel_attention(query, key, value)
+
+
+cp_function = local_map(
```
dead code? (But i think the block above is the real deal and the test is ok)

xmfan replied:
im gonna test the dynamo mapping for both

wconstab commented on test/higher_order_ops/test_local_map.py:
```diff
@@ -0,0 +1,123 @@
+    return out
+
+
+@local_map(
```
so IIUC the overall UX for this is that users use local_map as usual, and it 'just works' with torch compile- but for the time being, due to speculate inputs issue, you have to specifically enable this `LocalMapWrappedHigherOrderVariable` ctx, right?

xmfan replied:
yes, the code in this PR is to have dynamo trace the local map into a subgraph instead of into the main graph. this gives other components an easier time grabbing the nodes related to the local_map as a single unit (subgraph call, and the metadata lives on the call_function node targeting the subgraph).

if the inputs are DTensors we run into the speculate inputs issue, makes no sense to release this outside of AP rn

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update on "[dynamo][hop] Introduce Local Map HOP"


Can't actually deploy it because of: https://github.com/pytorch/pytorch/issues/161456

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

xmfan committed:
Update on "[dynamo][hop] Introduce Local Map HOP"


Can't actually deploy it because of: https://github.com/pytorch/pytorch/issues/161456

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

xmfan committed:
Update on "[dynamo][hop] Introduce Local Map HOP"


Can't actually deploy it because of: https://github.com/pytorch/pytorch/issues/161456

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

ydwu4 reviewed (approved):
the create_hop_fw_bw is a bit messy but we can clean it up later.

xmfan committed:
Update on "[dynamo][hop] Introduce Local Map HOP"


Can't actually deploy it because of: https://github.com/pytorch/pytorch/issues/161456

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

xmfan committed:
Update on "[dynamo][hop] Introduce Local Map HOP"


Can't actually deploy it because of: https://github.com/pytorch/pytorch/issues/161456

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

xmfan committed:
Update on "[dynamo][hop] Introduce Local Map HOP"


Can't actually deploy it because of: https://github.com/pytorch/pytorch/issues/161456

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

xmfan committed:
Update on "[dynamo][hop] Introduce Local Map HOP"


Can't actually deploy it because of: https://github.com/pytorch/pytorch/issues/161456

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]
